### DISCUSSION TITLE
Using CID to reference attached images in html mails

### DISCUSSION THREAD

**Discussion by kolaente** [wneessen/go-mail#27] [Q&A] [answered]
Hey there, really nice library, coming from gomail.

I've been trying to embed an image in a mail and referencing it via `cid:` but couldn't really get it to work.

Basically, I want to do this: https://stackoverflow.com/a/41994121/10924593

I've managed to add the file to the mail as an attachment but using `cid:filename.png` to actually reference the image does not seem to work.
I think I'd need to set an extra attachment header to set the content id? Or something like that?

**Comment by kolaente** [comment, marked as answer]
Okay looks like this just works™ when I use `EmbedReader` instead of manually adding the file as an attachment.

**Reply by wneessen** [reply]
> Hey there, really nice library, coming from gomail.
> 
Thanks! Glad you like it and happy if it provides you with value!

> I've been trying to embed an image in a mail and referencing it via `cid:` but couldn't really get it to work.
> Basically, I want to do this: https://stackoverflow.com/a/41994121/10924593
> 
As you already noticed yourself, the `Msg.Embed*()` methods are the way to go for inlining images and referencing them via "CID" tag. I've actually did some tests and Outlook seems to be happy with either way - attaching or embeding. Thunderbird and Webmail clients instead do only show this correctly if the file is inlined. Therefore I suggest to always perfer those methods. 

You don't have to use the `EmbedReader` btw. You can also use `Msg.EmbedFile()` directly.

**Reply by wneessen** [reply]
Maybe as an additional side note on why this happens...

The Embed-methods add a `inline` tag to the `Content-Disposition` header, as well as a `Content-Id: <file.xxx>` header to the attached file headers. The `Content-ID` header is what is referred to when using the `CID` tag.

**Reply by kolaente** [reply]
Awesome, thanks for the explanation.

> You don't have to use the `EmbedReader` btw. You can also use `Msg.EmbedFile()` directly.

I know, but using the Reader allows me to embed the file with `go:embed`. Or can I do that with `EmbedFile` as well?

**Reply by wneessen** [reply]
No, you are correct. At the moment. for go:embed the `EmbedReader()` is the right solution. I might add a `EmbedFile` version that supports embed.FS in one of the next releases.

**Reply by wneessen** [reply]
With the latest release: https://github.com/wneessen/go-mail/releases/tag/v0.2.5 you can now embed files from `embed.FS` directly without having to use `EmbedReader()`

**Reply by kolaente** [reply]
Works perfectly! Thanks a lot!
